cpp/arcticdb/version/version_core.cpp

```diff
@@ -304,92 +324,137 @@ void check_update_data_is_sorted(InputTensorFrame& frame, const index::IndexSegm
+                    index_segment_reader=std::move(index_segment_reader),
+                    frame,
+                    dynamic_schema,
+                    update_info,
+                    store](IntersectingSegments&& intersecting_segments) {
+                size_t row_count = 0;
+                const size_t new_keys_size = new_slice_and_keys.size();
+                const std::array<std::vector<SliceAndKey>, 5> groups{
+                    strictly_before(update_ranges.original, unaffected_keys),
+                    std::move(std::get<0>(intersecting_segments)),
+                    std::move(new_slice_and_keys),
+                    std::move(std::get<1>(intersecting_segments)),
+                    strictly_after(update_ranges.original, unaffected_keys)};
+                auto flattened_slice_and_keys = flatten_and_fix_rows(groups, row_count);
+
+                util::check(unaffected_keys.size() + new_keys_size + (affected_keys.size() * 2) >= flattened_slice_and_keys.size(),
+                            "Output size mismatch: {} + {} + (2 * {}) < {}",
+                            unaffected_keys.size(), new_keys_size, affected_keys.size(), flattened_slice_and_keys.size());
+                std::sort(std::begin(flattened_slice_and_keys), std::end(flattened_slice_and_keys));
+                auto tsd = index::get_merged_tsd(row_count, dynamic_schema, index_segment_reader.tsd(), frame);
+                return index::write_index(
+                    index_type_from_descriptor(tsd.as_stream_descriptor()),
+                    std::move(tsd),
+                    std::move(flattened_slice_and_keys),
+                    IndexPartialKey{frame->desc.id(), update_info.next_version_id_}, store
+                );
+            });
+        });
+    });
+}
 
-    std::sort(std::begin(flattened_slice_and_keys), std::end(flattened_slice_and_keys));
-    auto tsd = index::get_merged_tsd(row_count, dynamic_schema, index_segment_reader.tsd(), frame);
-    auto version_key_fut = index::write_index(stream::index_type_from_descriptor(tsd.as_stream_descriptor()), std::move(tsd), std::move(flattened_slice_and_keys), IndexPartialKey{stream_id, update_info.next_version_id_}, store);
-    auto version_key = std::move(version_key_fut).get();
+VersionedItem update_impl(
+    const std::shared_ptr<Store>& store,
+    const UpdateInfo& update_info,
+    const UpdateQuery& query,
+    const std::shared_ptr<InputTensorFrame>& frame,
+    WriteOptions&& options,
+    bool dynamic_schema,
+    bool empty_types) {
+    auto version_key = async_update_impl(store, update_info, query, frame, std::move(options), dynamic_schema, empty_types).get();
     auto versioned_item = VersionedItem(to_atom(std::move(version_key)));
-    ARCTICDB_DEBUG(log::version(), "updated stream_id: {} , version_id: {}", stream_id, update_info.next_version_id_);
+    ARCTICDB_DEBUG(log::version(), "updated stream_id: {} , version_id: {}", frame->desc.id(), update_info.next_version_id_);
```

Comment:
Is this set on the input frame? If not, can use the stream ID from the `version_key`